Implement default transformation flags for gpfdist

InnerLife0 opened the pull request:
Implement default transformation flags for gpfdist

Currently, gpfdist transformation can only be applied to table source on
external table level. It can be done by specifying #transform tag. Sometimes it
may be not handy to specify the same transformation for all source files. The
goal of the patch is to implement new flags with the default transformation name
for input and output transformations. The default transformation(s) can be set
at gpfdist start and applied only if table level transformation is not
specified.

New flags are a subject of only basic check. The idea was to keep the same
behavior and check transformation existence only at runtime.
The test shows a behavior described in patch goal.

_______
It seems that gpfdist has never been a target of pgindent utility. I prefered to keep the same code style.

BenderArenadata commented:
Allure report https://allure.adsw.io/launch/68774

BenderArenadata commented:
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1269451

BenderArenadata commented:
Failed job Regression tests with ORCA on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1269444

BenderArenadata commented:
Failed job Regression tests with ORCA on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1269445

Commit:
Implement default transformation flag for gpfdist

Currently, gpfdist transformation can only be applied to table source on
external table level. It can be done by specifying #transform tag. Sometimes it
may be not handy to specify the same transformation for all source files. The
goal of the patch is to implement a new flag with the default transformation
name. The default transformation can be set at gpfdist start and applied only if
table level rule is not specified.

New flag is a subject of only basic check. The idea was to keep the same
behavior and check rule existence only at runtime.
The test shows a behavior described in patch goal. Writable tables are not
considered here as their behavior is the same.

BenderArenadata commented:
Allure report https://allure.adsw.io/launch/68832

BenderArenadata commented:
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1272018

BenderArenadata commented:
Failed job Regression tests with Postgres on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1272009

bandetto commented on src/bin/gpfdist/gpfdist.c:
```diff
@@ -175,6 +176,7 @@ static struct
 	int			w; /* The time waiting when used for session timeout in seconds */
 	int 		k; /* The time used to clean up sessions in seconds */
+	const char* X; /* default transformation */
 } opt = { 8080, 8080, 0, 0, 0, ".", 0, 0, -1, 5, 0, 32768, 0, 256, 0, 0, 0, 0 , 300};
```
X isn't initialized to anything.
```suggestion
} opt = { 8080, 8080, 0, 0, 0, ".", 0, 0, -1, 5, 0, 32768, 0, 256, 0, 0, 0, 0 , 300, NULL};
```

InnerLife0 replied:
Thanks. But to keep the same code style it should be 0, not NULL.

bandetto commented on src/bin/gpfdist/gpfdist.c:
```diff
@@ -533,6 +535,7 @@ static void usage_error(const char* msg, int print_usage)
 #endif
 #ifdef GPFXDIST
 					    "        -c file    : configuration file for transformations\n"
+					    "        -X name    : name of default transformation\n"
```
`gpfdist --help` takes help from this file in case `$GPHOME` is available.
https://github.com/arenadata/gpdb/blob/4003b8f0fc15476e7bb19b667cebac269f8a7867/src/bin/gpfdist/gpfdist.c#L495-L501
It's source is located in `gpMgmt/doc/gpfdist_help`. I think it should be updated too.

InnerLife0 replied:
True. Fixed.

bandetto replied:
Flags are shown in the synopsis too:
https://github.com/arenadata/gpdb/blob/0dc96e8dca4c27aacde659f7f7f7bcb2516e7fc2/gpMgmt/doc/gpfdist_help#L10-L12
How about adding `[-X <default transformation>]` there as well?

InnerLife0 replied:
Done

bandetto commented on src/bin/gpfdist/regress/input/exttab1.source:
```diff
@@ -650,6 +650,44 @@ SELECT * FROM readable_external_table_with_user_type_and_default_value;
+-- The server runs with default transformation of catfile_double which doubles
+-- file content
+CREATE EXTERNAL WEB TABLE gpfdist_transform_start (x text)
+execute E'((@bindir@/gpfdist -p 7070 -d @abs_srcdir@/data/ -c @abs_srcdir@/data/catfile.yml -X catfile_double </dev/null >/dev/null 2>&1 &); for i in `seq 1 30`; do curl 127.0.0.1:7070 >/dev/null 2>&1 && break; sleep 1; done; echo "starting...") '
```
What's the purpose of `echo "starting..."`?

InnerLife0 replied:
Just a dumb log message (or a comment) about a start of gpfdist server. The same like it's done [here](https://github.com/arenadata/gpdb/blob/4003b8f0fc15476e7bb19b667cebac269f8a7867/src/bin/gpfdist/regress/input/exttab1.source#L41) or in any other example in file. I don't think it's worth to change something here.

Commit:
Add def value for X. Update gpfdist_help.

BenderArenadata commented:
Allure report https://allure.adsw.io/launch/69099

BenderArenadata commented:
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1285772

BenderArenadata commented:
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1285773

Commit:
Add -X to synopsys

BenderArenadata commented:
Allure report https://allure.adsw.io/launch/69131

BenderArenadata commented:
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1286903

BenderArenadata commented:
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1286904

whitehawk commented on src/bin/gpfdist/gpfdist.c:
```diff
@@ -3535,6 +3565,9 @@ static int request_set_transform(request_t *r)
 	}
 
+	if (! r->trans.name && opt.X)
+		r->trans.name = opt.X;
```
What if:
(1) there is a transformation with name SOME_TRANSFORMATION and type=input.
(2) SOME_TRANSFORMATION is set to be default transformation for gpfdist.
(3) there is a WRITABLE external table without a transformation.

Will not writing to this table result in error here? As `transform_lookup` will fail to find SOME_TRANSFORMATION and type=output...
```
    /*
     * at this point r->trans.name is the name of the transformation requested
     * in the url and r->is_get tells us what kind (input or output) to look for.
     * attempt to look it up.
     */
	tr = transform_lookup(opt.trlist, r->trans.name, r->is_get ? 0 : 1, opt.V);
    if (! tr)
    {
        if (r->is_get)
        {
            gprintln(r, "reject invalid request from %s [%s %s] - unsppported input #transform",
                     r->peer, r->in.req->argv[0], r->in.req->argv[1]);
            http_error(r, FDIST_BAD_REQUEST, "invalid request (unsupported input #transform)");
        }
        else
        {
            gprintln(r, "reject invalid request from %s [%s %s] - unsppported output #transform",
                     r->peer, r->in.req->argv[0], r->in.req->argv[1]);
            http_error(r, FDIST_BAD_REQUEST, "invalid request (unsupported output #transform)");
        }
        request_end(r, 1, 0);
        return -1;
    }
```

BTW, there is a typo in "unsppported".

InnerLife0 replied:
This case is an expected behavior. We act the same as if transformation is specified by user on table level. User will get the same error if transformation is not found.
Not related, but interesting case, is that file can contain input and output rule with the same name.

As about a typo, we shouldn't fix this in current patch.

whitehawk replied:
Indeed, current implementation meets the requirements in the initial technical proposal.

But, thinking more widely:
The original request from the customer was related to input transformations. Because they had problems when loading data into GPDB. As far as I understand, they didn't want to have a default tranformation for WRITABLE external tables.

Now, with this change, I think there is a possibility that
 - at first launch customer will set up only 'input' default transformation.
 - after that the customer will face that work with WRITABLE tables (that have no transformation specified) is broken.
 - with our support, or by himself, the customer will realize that he needs to setup a dummy 'output' default tranformation to make WRITABLE tables working. And that will look like an unnecessary W/A (as he had no issues with writable tables at all).

I'm worried, that during this route, the customer may become a bit upset.

So, if we are staying with current approach for same default transormation for both input and output, let's at least highlight this aspect in the helpfile.

And please discuss this issue with the architect - maybe the technical proposal needs adjustment.

whitehawk commented on src/bin/gpfdist/gpfdist.c:
```diff
@@ -836,6 +843,29 @@ static void parse_command_line(int argc, const char* const argv[],
+				" but TRANSFORMATIONS are not specified", 0);
+
+		p = gstring_trim(apr_pstrdup(pool, opt.X));
+		p_len = strlen(p);
```
For other options the value allocated in apr_pstrdup (variable `p`) is assigned to `opt`, for ex:

```
		opt.d = p;
```
and 
```
		opt.l = p;
```
and
```
		opt.ssl = p;
```

why the `p` is not assigned to `opt.X`?

InnerLife0 replied:
That was a mistake. Thanks.

Commit:
Use normalized value as X

BenderArenadata commented:
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1298447

BenderArenadata commented:
Failed job Regression tests with Postgres on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1298436

Commit:
Implement default transformation flags for gpfdist

Currently, gpfdist transformation can only be applied to table source on
external table level. It can be done by specifying #transform tag. Sometimes it
may be not handy to specify the same transformation for all source files. The
goal of the patch is to implement new flags with the default transformation name
for input and output transformations. The default transformation(s) can be set
at gpfdist start and applied only if table level transformation is not
specified.

New flags are a subject of only basic check. The idea was to keep the same
behavior and check transformation existence only at runtime.
The test shows a behavior described in patch goal.

Commit:
Merge branch 'adb-6.x-dev' into ADBDEV-5344

Commit:
Uncomment production code

Commit:
Remove debug line

Commit:
Fix comments in test

whitehawk commented:
In the summary "Implement default transformation flag for gpfdist" please change "flag" to "flags".

whitehawk commented on src/bin/gpfdist/transform.c:
```diff
@@ -781,4 +781,31 @@ transform_saferegex(struct transform* tr)
+	 * transformation to all transformations is checked at runtime.
+	 */
+	if (name_len < 1 || name_len > MAX_KEYLEN)
+		return "Error: -X default transformation name length must be"
```
I guess the `-X` here is a rudiment of the previous patch version and the error message should also be updated to use `-I` and `-O` instead of `-X`.

Getting `-X` in error message might confuse user, as there is no such option now.

InnerLife0 replied:
Fixed

whitehawk commented on src/bin/gpfdist/regress/input/exttab1.source:
```diff
@@ -650,6 +650,84 @@ SELECT * FROM readable_external_table_with_user_type_and_default_value;
+FORMAT 'text' (delimiter '|');
+
+CREATE EXTERNAL WEB TABLE gpfdist_transform_start2 (x text)
+execute E'((@bindir@/gpfdist -p 7080 -d @abs_srcdir@/data/ -c @abs_srcdir@/data/catfile.yml -O catfile_double </dev/null >/dev/null 2>&1 &); for i in `seq 1 30`; do curl 127.0.0.1:7070 >/dev/null 2>&1 && break; sleep 1; done; echo "starting...") '
```
Is it correct here to access port 7070 in `do curl 127.0.0.1:7070`, while gpfdist was started on 7080? Typo?

InnerLife0 replied:
Yes, it's a copied typo. Fixed.

Commit:
Fix port and error message

BenderArenadata commented:
Allure report https://allure.adsw.io/launch/69572

BenderArenadata commented:
Failed job Behave tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1314714

Commit:
Merge branch 'adb-6.x-dev' into ADBDEV-5344

andr-sokolov commented on src/bin/gpfdist/gpfdist.c:
```diff
@@ -836,6 +848,37 @@ static void parse_command_line(int argc, const char* const argv[],
+	/* validate opt.I and opt.O */
+	if (opt.I || opt.O)
+	{
+		extern const char* validate_transform_opt(struct transform* trlist, const char* name);
```
Can we include transform.h instead of this extern statement? When someday arguments of the validate_transform_opt function will be changed, a compiler won't be able to detect the error.

InnerLife0 replied:
You are TOTALLY right. But for some historical reasons, gpfdist developers decided to do it like so (there are many examples of it in the same file). Doing it another way is a refactoring, which I don't want to include in this patch.

BenderArenadata commented:
Allure report https://allure.adsw.io/launch/69927

andr-sokolov commented on src/bin/gpfdist/gpfdist.c:
```diff
@@ -836,6 +848,37 @@ static void parse_command_line(int argc, const char* const argv[],
     }
+
+	/* validate opt.I and opt.O */
+	if (opt.I || opt.O)
```
can we delete this `if` statment to don't check opt.I and opt.O twince?
```c
    extern const char* validate_transform_opt(struct transform* trlist, const char* name);
	if (opt.I)
	{
		char *p = gstring_trim(apr_pstrdup(pool, opt.I));
		const char *err = validate_transform_opt(opt.trlist, p);

		if (err)
			usage_error(err, 0);

		opt.I = p;
	}

	if (opt.O)
	{
		char *p = gstring_trim(apr_pstrdup(pool, opt.O));
		const char *err = validate_transform_opt(opt.trlist, p);

		if (err)
			usage_error(err, 0);

		opt.O = p;
	}
```

InnerLife0 replied:
In this example extern function is always defined, even if it's not used later.
Another reason was to avoid, not all, but a bit of copypaste.

Anyway, I think it's like "exchange an awl for soap" here.

andr-sokolov replied:
I think than there is no problem when a function is defined more global. If we have less checks, the code become simpler. And some microseconds faster)

InnerLife0 replied:
This is more not about a check, but about code readability. Placing `extern` inside if condition informs the developer, that this function is used only in this block. The code above has example of the same situation. If if we put transform_config() out of if condition, then developer can't say that function is used only in next condition and should scroll below (to our new flags, or even lower to future changes) to check it.

From another points, it's awl for soap.

Commit:
Merge branch 'adb-6.x-dev' into ADBDEV-5344

BenderArenadata commented:
Allure report https://allure.adsw.io/launch/69995
